### PR TITLE
Add support for FlexAttention higher-order operator

### PR DESCRIPTION
**Pull request by fmassa**
FlexAttention (`torch.nn.attention.flex_attention`) is a higher-order operator (HOP) that appears in many PyTorch models. Previously, AutoParallel would crash when encountering it because the HOP's GraphModule submodules (score_mod, mask_mod) lack tensor metadata, and the HOP target isn't an OpOverload that the DTensor strategy system expects.

This PR treats `flex_attention` as an opaque box — like `local_map` — and enumerates valid input/output shardings based on attention semantics. Attention is independent per `(batch, head)` pair, so we allow `{Replicate, Shard(0), Shard(1)}` per mesh dimension. Block-mask and auxiliary tensors get per-dim placement adjusted to their actual shape: a `block_mask` created with `B=None` gets `Replicate` on the batch dim but can still be `Shard(1)` on heads if H is set.

The changes are best reviewed starting from `placement_options.py` (the new handler), then `optimize_sharding.py` (making the ILP robust to HOP submodule nodes), then `apply_sharding.py` (redistribution filtering for HOPs), and finally the tests.

Tested with `score_mod`, `block_mask` (broadcast and explicit B/H), custom scale, GQA (different Q vs KV head counts), and combinations thereof, on both 1D and 2D meshes.

Co-authored-by: Claude

### CODE REVIEW

**Comment by fmassa** [comment]
Test failures are the same as in https://github.com/meta-pytorch/autoparallel/pull/419 , so they are unrelated to this PR.

Also cc @sanketpurandare as the failures are probably in the graph pp runner

File: tests/test_flex_attention.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+        _flatten_out_shardings,
+        _make_input_fn,
+    )
+
+    x = _make_input(mesh, DIM)
+    shapes, dtypes, input_placements, treespec = _extract_input_info((x,), mesh)
+    input_fn = _make_input_fn(shapes, dtypes, treespec)
+    output_placements = _flatten_out_shardings(_out_shardings(mesh))
+
+    with AutoParallel(model, input_fn, mesh, compile=False) as autop:
```
Comment (IvanKobzarev) [inline tests/test_flex_attention.py:482]: No tests with compile=True as it can be slow on CI ?
At least one compilation will be good to check I think.

File: autoparallel/shardings/placement_options.py [hunk 2]
```diff
@@ -483,12 +484,165 @@ def get_local_map_placement_option(
...
+    def tensor_placement(t, placement):
+        """Compute per-tensor placement, replacing Shard dims that don't match
+        the reference size with Replicate (e.g. block_mask with B=1)."""
+        dim_to_ref = {0: B, 1: H}
+        adjusted = []
+        for p in placement:
+            if (
+                p.is_shard()
+                and p.dim in dim_to_ref
+                and t.shape[p.dim] != dim_to_ref[p.dim]
```
Comment (IvanKobzarev) [inline autoparallel/shardings/placement_options.py]: In case of GQA we will have Q 64 heads and KV 8 heads. (Llama3)
With check != it will always make it Replicate.
Could it be `dim_to_ref[p.dim] % t.shape[p.dim] != 0` to allow GQA to use sharding on head_dim?

**Review by IvanKobzarev — APPROVED** [review]

File: tests/test_flex_attention.py (hunk 1 above)
Comment (fmassa) [reply tests/test_flex_attention.py:482]: Will add a test with `compile=True`, good catch!

File: autoparallel/shardings/placement_options.py (hunk 2 above)
Comment (fmassa) [reply autoparallel/shardings/placement_options.py]: Done, thanks for the catch!
